step01/main.py

```diff
@@ -0,0 +1,175 @@
+    receipt_handle = message['ReceiptHandle']
+
+    return cls(_id, key, receipt_handle)
+
+
+def gen_messages_from_response(response):
+  '''
+  This function parses a response from AWS ReceiveMessages and returns
+  a generator that yields one message at a time for each message in the response.
+
+  Gets only one message: MaxNumberOfMessages = 1
+  Long Polling: WaitTimeSeconds = 2
+  '''
+  raw_messages = response.get('Messages', [])
+  for raw_message in raw_messages:
+    yield S3Message.parse(raw_message)
+
+
+def create_thumbnail(input_stream, size=(128, 128)):
+  '''
+  This function creates a thumbnail of an image. It expects bytes and returns
+  bytes.
+  '''
+
+  output_stream = BytesIO()
+  image = Image.open(BytesIO(input_stream))
+  image.thumbnail(size)
+  image.save(output_stream, image.format)
+  output_stream.seek(0)
+  return output_stream
+
+
+def create_thumbnail_key(key):
+  '''
+  This function creates the S3 key for the thumbnail to be uploaded to S3
+  based on the key of the original image.
+  '''
+
+  key_path = PurePath(key)
+  file_stem = key_path.stem
+  file_ext = key_path.suffix
+  return '{}/{}.thumbnail{}'.format(OUTPUT_FOLDER, file_stem, file_ext)
+
+
+def main():
+  print('Waiting for Messages...')
+  while True:
+    response = sqs.receive_message(
+      QueueUrl=QUEUE_URL,
+      MaxNumberOfMessages=1,
+      WaitTimeSeconds=2
+    )
+
+    messages = gen_messages_from_response(response)
+    for message in messages:
+      try:
+        get_object_response = s3.get_object(
+          Bucket=BUCKET_NAME,
+          Key=message.key
+        )
+
+        print(f'Creating Thumbnail for {message._id}')
+        stream = get_object_response['Body'].read()
+        thumbnail_stream = create_thumbnail(stream)
+
+        print(f'Uploading Thumbnail for {message._id}')
+        thumbnail_key = create_thumbnail_key(message.key)
+        s3.put_object(
+          Bucket=BUCKET_NAME,
+          Key=thumbnail_key,
+          Body=thumbnail_stream
+        )
+
+        print(f'Deleting Message {message._id}')
+        sqs.delete_message(
+          QueueUrl=QUEUE_URL,
+          ReceiptHandle=message.receipt_handle
+        )
+
+      except ClientError as botoError:
+        '''        
+        ClientError
+        - boto3 errors (s3, sqs)
+        '''
+        print(f'Botocore Error: {botoError}')
+        pass
+
+      except KeyError:
+        '''error from no output format'''
+        print(f'Output error: {KeyError}')
+        pass
+
+      except IOError as IOe:
+        '''  
+        - errors from file not found or opened
+        - errors if file not written
+        
+        Note: most errors from an image file or supported image file
+        '''
+        print(f'IO Error: {IOe}')
+        print('Thumbnail not created.')
+        print(f'Deleting Message {message._id}')
+        # Deletes message to prevent further processing
+        sqs.delete_message(
```

Comment:
You do not need to change anything here, but I want you to think about what would happen if any of `sqs.delete_message` or `s3.delete_message` fails. You should see that having a try-catch here would produce lots of duplicate code.

Reply:
FYI: assuming you meant `s3.delete_object`

Is this one of those times where you should not care about it failing?

In my mind if either fails, neither appear to cause anything major damage. Maybe:
- same exception runs if message is processed again
- object remains in s3 bucket 

I can see how a chasing all the exceptions using try-except could end up being an on going battle with duplicate code as you said. In this case, maybe some kind of logging could be used?

Reply:
This is definitely something you would want to handle. The particular outcome you'd want to avoid is to have the same data errors resulting in different outcomes depending on whether another error occurred. You will end up with messages building up in the queue but have a hard time figuring out what's causing it. 

We are skipping this because you'll be dealing with it in a later challenge

Reply:
Okay, understood. Will keep in mind for the future.